quickfort.lua

```diff
@@ -185,6 +185,157 @@ local function do_set(args)
+example: '#dig (start 4;4;center of stairs) dining hall'
+where all elements other than the initial #mode are optional (though if the
+'start' block exists, the offsets must also exist)
+returns a table in the format {mode, startx, starty, start_comment, comment}
+or nil if the modeline is invalid
+]]
+local function parse_modeline(line)
+    if not line then return nil end
+    local _, mode_end, mode = string.find(line, '^#([%l]+)')
+    if not mode or not mode_switch[mode] then
+        print(string.format('invalid mode: %s', mode))
+        return nil
+    end
+    local _, start_str_end, start_str = string.find(
+        line, '%s+start(%b())', mode_end + 1)
+    local startx, starty, start_comment = 1, 1, nil
+    if start_str then
+        _, _, startx, starty, start_comment = string.find(
+            start_str, '^%(%s*(%d+)%s*;%s*(%d+)%s*;?%s*(.*)%)$')
+        if not startx or not starty then
+            print(string.format('invalid start offsets: %s', start_str))
+            return nil
+        end
+    else
+        start_str_end = mode_end
+    end
+    local _, _, comment = string.find(line, '%s*(.*)', start_str_end + 1)
+    return {
+        mode=mode,
+        startx=startx,
+        starty=starty,
+        start_comment=start_comment,
+        comment=comment
+    }
+end
+
+local function get_modeline(filepath)
+    local file = io.open(filepath)
+    local first_line = file:read()
+    file:close()
+    if (not first_line) then return nil end
+    return parse_modeline(tokenize_csv_line(first_line)[1])
+end
+
+-- filename is relative to the blueprints dir
+local function get_blueprint_filepath(filename)
+    return string.format("%s/%s",
+                         quickfort_common.settings['blueprints_dir'], filename)
+end
+
+local blueprint_cache = {}
+
+local function scan_blueprint(path)
+    local filepath = get_blueprint_filepath(path)
+    local hash = dfhack.internal.md5File(filepath)
+    if not blueprint_cache[path] or blueprint_cache[path].hash ~= hash then
+        blueprint_cache[path] = {modeline=get_modeline(filepath), hash=hash}
+    end
+    return blueprint_cache[path].modeline
+end
+
+local blueprint_files = {}
+
+local function scan_blueprints()
+    local paths = dfhack.filesystem.listdir_recursive(
+        quickfort_common.settings['blueprints_dir'], nil, false)
+    blueprint_files = {}
+    local library_files = {}
+    for _, v in ipairs(paths) do
+        if not v.isdir and
+                (string.find(v.path, '[.]csv$') or
+                 string.find(v.path, '[.]xlsx$')) then
+            if string.find(v.path, '[.]xlsx$') then
+                print(string.format(
+                        'skipping "%s": .xlsx files not supported yet', v.path))
+                goto skip
+            end
+            local modeline = scan_blueprint(v.path)
+            if not modeline then
+                print(string.format(
+                        'skipping "%s": no #mode marker detected', v.path))
```

Comment:
Do you have plans to support these later? I happened to have a very bare-bones CSV file I was testing digfort with yesterday, just containing a couple rows and columns:
```
d,h,d
h,d,h
d,h,d
```
I don't know how widespread files like these are, if at all (and I'm fine with not supporting them)

Reply:
Could you test to see if it works in the original quickfort windows app? I *could* support it, and just assume that no header means #dig -- I don't *think* that's something that will bite us later..

At some point I'm thinking of supporting multiple "sheets" in the same .csv file, which would give them a lot of the advantages that .xlsx files have. each "sheet" would be separated by a modeline, but I don't think supporting missing initial modelines would interfere with that -- it's just subsequent "sheets" that would need one.

I'll code it up

Reply:
we risk polluting the 'quickfort list' output with unrelated files, but how likely is it that users are storing unrelated .csv or .xlsx files in the blueprints folder?

I guess a greater risk is when we support .xlsx, users can't have non-blueprint sheets in their spreadsheet files. I myself usually keep a "Notes" sheet in my spreadsheet files. we could only support headerless .csv files and not headerless .xlsx sheets, but that might be confusing for users who convert their working .csv files into non-working .xlsx files

Reply:
I don't have ready access to Windows, sorry. All I know is that digfort accepts it. I suspect that people wouldn't have non-blueprints in the `blueprints` folder, particularly since it's new (to DFHack).

Multiple blueprints in a single file isn't something I had considered... if that makes it complicated to handle this, I'm fine with putting this off.

Reply:
so I guess to prevent potential confusion later, I think I'd rather not implement modeless blueprints -- unless, of course, there is a clear signal from the community that this is wanted. I'm not seeing a lot of activity in the thread I started, so I'm not sure it's even worth asking now, but I expect more activity once this script gets into a release and starts seeing some real usage.

Reply:
Sounds good to me!